### PR TITLE
Remove ValidateSecurity layer of abstraction

### PR DESCRIPTION
**Pull request by orenbm**
As part of the authentication process we validate that the authenticator
webservice is whitelisted and can be accessed by the requesting role.
These validations were encapsulated by ValidateSecurity but it made
the code less clear when reading the `call` methods of the command-classes
that were calling this validation.

Removing this layer, and calling the validations themselves will enhance the
readability of the authentication classes which are some of the main parts
 of the product.

### CODE REVIEW

File: spec/app/domain/authentication/authenticate_spec.rb [hunk 1]
```diff
@@ -94,8 +75,11 @@ def authenticator(pass:)
       )
 
       Authentication::Authenticate.new(
-        token_factory: token_factory,
-        audit_log: audit_logger
+        validate_webservice_access:      mock_validate_webservice_access(validation_succeeded: true),
+        validate_whitelisted_webservice: mock_validate_whitelisted_webservice(validation_succeeded: true),
+        validate_origin:                 mocked_origin_validator,
+        token_factory:                   token_factory,
```
Comment (Tovli) [inline spec/app/domain/authentication/authenticate_spec.rb:81]: The token factory is not mocked in this case?

File: app/domain/authentication/authenticate.rb [hunk 2]
```diff
@@ -27,7 +29,8 @@ module Authentication
 
     def call
       validate_authenticator_exists
-      validate_security
+      validate_webservice_is_whitelisted
+      validate_user_has_access_to_webservice
```
[the remark this answers is not in the document]
Comment (orenbm) [reply app/domain/authentication/authenticate.rb:31]: changed, thanks

File: spec/app/domain/authentication/authenticate_spec.rb (hunk 1 above)
Comment (orenbm) [reply spec/app/domain/authentication/authenticate_spec.rb]: renamed

File: app/domain/authentication/authenticate.rb [hunk 3]
```diff
@@ -11,10 +10,11 @@ module Authentication
 
   Authenticate ||= CommandClass.new(
     dependencies: {
-      token_factory:     TokenFactory.new,
-      validate_security: ::Authentication::Security::ValidateSecurity.new,
-      validate_origin:   ::Authentication::ValidateOrigin.new,
-      audit_log:         ::Audit.logger
+      token_factory:                       TokenFactory.new,
+      validate_webservice_is_whitelisted:  ::Authentication::Security::ValidateWebserviceIsWhitelisted.new,
+      validate_user_can_access_webservice: ::Authentication::Security::ValidateUserCanAccessWebservice.new,
```
Comment (Tovli) [inline app/domain/authentication/authenticate.rb:15]: The word "user" is right here? are't hosts use authn?
or it's the same for both users and hosts?

File: app/domain/authentication/authenticate.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sigalsax) [reply app/domain/authentication/authenticate.rb:31]: couldn't hosts also ideally authenticate to the webservice? if so, consider `validate_role_has_access_to_webservice`

File: app/domain/authentication/security/validate_webservice_exists.rb [hunk 4]
```diff
@@ -1,12 +1,9 @@
...
 
-require 'authentication/validate_account_exists'
-
 module Authentication
 
   module Security
 
-    Err ||= Errors::Authentication::Security
     # Possible Errors Raised:
     # AccountNotDefined, WebserviceNotFound
```
Comment (sigalsax) [inline app/domain/authentication/security/validate_webservice_exists.rb:8]: Doesn't seem like `AccountNotDefined`  is used

Comment (sigalsax) [reply app/domain/authentication/security/validate_webservice_exists.rb:8]: Neither is `WebserviceNotFound`

File: spec/app/domain/authentication/authenticate_spec.rb [hunk 5]
```diff
@@ -135,64 +124,123 @@ def authenticator(pass:)
...
-                                )
+        it "returns a new token" do
+          expect(subject).to(
+            equal(
+              a_new_token
+            )
+          )
         end
+
+        it_behaves_like "raises an error when origin validation fails"
```
Comment (sigalsax) [inline spec/app/domain/authentication/authenticate_spec.rb:243]: Why is this here? The context is `when webservice validations succeed` so how does that relate to an request coming from a blocked origin?

File: spec/app/domain/authentication/authn-oidc/authn_oidc_spec.rb [hunk 6]
```diff
@@ -85,6 +85,51 @@ def request_body(request)
...
+            ).call(
+              authenticator_input: input_
+            )
+          end
+
+          it "returns a new access token" do
+            expect(subject).to equal(a_new_token)
+          end
+
+          it_behaves_like "raises an error when origin validation fails"
```
Comment (sigalsax) [inline spec/app/domain/authentication/authn-oidc/authn_oidc_spec.rb:118]: Seems like the `it_behaves_like` list below are separate tests? Is `it_behaves_like` just `context` but nested?

File: spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb [hunk 7]
```diff
@@ -70,43 +70,88 @@
...
-      allow(k8s_object_lookup_class).to receive(:pod_by_name)
-                                          .with(spiffe_name, spiffe_namespace)
-                                          .and_return(nil)
+      it 'raises an error when application identity validation fails' do
+        allow(validate_application_identity).to receive(:call)
+                                                  .and_raise('FAKE_APPLICATION_IDENTITY_ERROR')
+
+        expect { subject }.to(
+          raise_error(
+            /FAKE_APPLICATION_IDENTITY_ERROR/
```
Comment (sigalsax) [inline spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb:105]: In a previous test, ` /FAKE_APPLICATION_IDENTITY_ERROR/` was removed and replaced with `validate_webservice_is_whitelisted_error` which is 
`let(:validate_webservice_is_whitelisted_error) { "validate whitelisted webservice error" }`

so IMO it will look cleaner if we  choose a style.

File: spec/support/security_specs_helper.rb [hunk 8]
```diff
@@ -106,29 +102,15 @@ def mock_role
...
   context 'when origin validation fails' do
     let(:audit_success) { false }
-    
+
     it "raises an error" do
       allow(mocked_origin_validator).to receive(:call)
                                           .and_raise('FAKE_ORIGIN_ERROR')
 
-      expect { subject }.to raise_error( /FAKE_ORIGIN_ERROR/ )
+      expect { subject }.to raise_error(/FAKE_ORIGIN_ERROR/)
```
Comment (sigalsax) [inline spec/support/security_specs_helper.rb:113]: Same comment here about choose a style of a mock or `FAKE_ORIGIN_ERROR`

File: app/domain/authentication/authenticate.rb (hunk 3 above)
Comment (orenbm) [reply app/domain/authentication/authenticate.rb]: changed to `role`

File: app/domain/authentication/security/validate_webservice_exists.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/security/validate_webservice_exists.rb:8]: removed

File: spec/app/domain/authentication/authenticate_spec.rb (hunk 5 above)
Comment (orenbm) [reply spec/app/domain/authentication/authenticate_spec.rb:243]: this is inside the vanilla flow, just like before. We verify that it fails when origin validation fails (and everything else succeeds)

File: spec/app/domain/authentication/authn-oidc/authn_oidc_spec.rb (hunk 6 above)
Comment (orenbm) [reply spec/app/domain/authentication/authn-oidc/authn_oidc_spec.rb:118]: yes, they are separate tests. you can see their definition in the spec helper

File: spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb (hunk 7 above)
Comment (orenbm) [reply spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb:105]: Done

File: spec/support/security_specs_helper.rb (hunk 8 above)
Comment (orenbm) [reply spec/support/security_specs_helper.rb:113]: Done

File: spec/app/domain/authentication/authenticate_spec.rb [hunk 9]
```diff
@@ -94,8 +75,11 @@ def authenticator(pass:)
       )
 
       Authentication::Authenticate.new(
-        token_factory: token_factory,
-        audit_log: audit_logger
+        validate_role_can_access_webservice: mock_validate_user_can_access_webservice(validation_succeeded: true),
```
Comment (sigalsax) [inline spec/app/domain/authentication/authenticate_spec.rb:78]: Because you changed `validate_user_can_access_webservice` -> `validate_role_can_access_webservice`, please follow suit here (:

File: app/domain/authentication/security/validate_account_exists.rb [hunk 10]
```diff
@@ -4,8 +4,6 @@ module Authentication
 
   module Security
 
-    Err ||= Errors::Authentication::Security
```
Comment (sigalsax) [inline app/domain/authentication/security/validate_account_exists.rb:7]: there are other places in the codebase that still have `Err ||=` . Was that intentional?

Comment (orenbm) [reply app/domain/authentication/security/validate_account_exists.rb:7]: i didn't address this in the current PR
